### PR TITLE
Enhanced interpolation

### PR DESCRIPTION
**Pull request by jodastephen**
Move interpolator/extrapolator interface to strata-market
Separate from strata-math interpolator/extrapolator
Remove serializable where possible
Design for potential future enhancements

### CODE REVIEW

File: modules/math/src/main/java/com/opengamma/strata/math/impl/interpolation/Interpolator1DFactory.java [hunk 1]
```diff
@@ -232,16 +229,16 @@ public static String getInterpolatorName(final Interpolator1D interpolator) {
     return s_instanceNames.get(interpolator.getClass());
   }
 
-  public static CurveInterpolator findCurveInterpolator(final String interpolatorName) {
+  public static Interpolator1D findCurveInterpolator(final String interpolatorName) {
```
Comment (cjkent) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/interpolation/Interpolator1DFactory.java]: The name of this method clashes with the convention that `find*` should return an `Optional` and `get*` should throw an exception if the data isn't found.

File: modules/market/src/main/java/com/opengamma/strata/market/interpolator/ImmutableCurveInterpolator.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  //-------------------------------------------------------------------------
+  @Override
+  public BoundCurveInterpolator bind(
+      DoubleArray xValues,
+      DoubleArray yValues,
+      CurveExtrapolator extrapolatorLeft,
+      CurveExtrapolator extrapolatorRight) {
+
+    if (!(extrapolatorLeft instanceof ImmutableCurveExtrapolator)) {
```
Comment (cjkent) [inline modules/market/src/main/java/com/opengamma/strata/market/interpolator/ImmutableCurveInterpolator.java]: Is this `instanceof` check intended to be temporary?

File: modules/math/src/main/java/com/opengamma/strata/math/impl/interpolation/DoubleQuadraticInterpolator1D.java [hunk 3]
```diff
@@ -29,13 +27,10 @@
 @BeanDefinition(style = "light", constructorScope = "public")
 public final class DoubleQuadraticInterpolator1D
     extends Interpolator1D
-    implements CurveInterpolator, ImmutableBean, Serializable {
```
Comment (cjkent) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/interpolation/DoubleQuadraticInterpolator1D.java]: Shouldn't this be `Serializable` given that it's a bean?

**Comment by cjkent** [comment]
That's everything

File: modules/math/src/main/java/com/opengamma/strata/math/impl/interpolation/DoubleQuadraticInterpolator1D.java (hunk 3 above)
Comment (jodastephen) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/interpolation/DoubleQuadraticInterpolator1D.java]: Not all beans need to be serializable. In this case, bean is being used just for convenience. All interpolator implementations no longer need to be serializable.

**Pull request merged**